### PR TITLE
Assignment9+10

### PR DESCRIPTION
**Pull request by mpennock5**

### CODE REVIEW

**Review by czamora7 — COMMENTED** [review]
Implement PingProcess' public Task<PingResult> RunTaskAsync(string hostNameOrAddress) ✔
First implement public void RunTaskAsync_Success() test method to test PingProcess.RunTaskAsync() using "localhost". ✔
Do NOT use async/await in this implementation. ✔

Implement PingProcess' async public Task<PingResult> RunAsync(string hostNameOrAddress) ✔
First implement the public void RunAsync_UsingTaskReturn_Success() test method to test PingProcess.RunAsync() using "localhost" without using async/await. ✔
Also implement the async public Task RunAsync_UsingTpl_Success() test method to test PingProcess.RunAsync() using "localhost" but this time DO using async/await. ✔

Add support for an optional cancellation token to the PingProcess.RunAsync() signature. ✔ Inside the PingProcess.RunAsync() invoke the token's ThrowIfCancellationRequested() method so an exception is thrown. ✔ Test that, when cancelled from the test method, the exception thrown is an AggregateException ❌ 
with a TaskCanceledException inner exception ❌ the method will throw an OperationCanceledException
using the test methods RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrapping ✔and RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrappingTaskCanceledException ✔ respectively.

Complete/fix AND test async public Task<PingResult> RunAsync(IEnumerable<string> hostNameOrAddresses, CancellationToken cancellationToken = default) which executes ping for and array of hostNameOrAddresses (which can all be "localhost") in parallel, adding synchronization if needed. ✔ NOTE:
The order of the items in the stdOutput is irrelevent and expected to be intermingled.
StdOutput must have all the ping output returned (no lines can be missing) even though intermingled. ❌ This needs to be tested for

Implement AND test public Task<int> RunLongRunningAsync(ProcessStartInfo startInfo, Action<string?>? progressOutput, Action<string?>? progressError, CancellationToken token) using Task.Factory.StartNew() and invoking RunProcessInternal with a TaskCreation value of TaskCreationOptions.LongRunning and a TaskScheduler value of TaskScheduler.Current. Returning a Task<PingResult> is also okay. NOTE: This method does NOT use Task.Run.
❌ the method header does not take the specified parameters and RunProcessInternal is not invoked

Extra Credit
Test and implement PingProcess.RunAsync(System.IProgress<T> progess) so that you can capture the output as it occurs rather than capturing the output only when the process finishes. ❌ not attempted

File: Assignment/Assignment.Tests/PingProcessTests.cs [hunk 1]
```diff
@@ -59,42 +61,68 @@ public void RunTaskAsync_Success()
...
+        Task<PingResult> task = Sut.RunAsync("localhost", token);
+
+        try
+        {
+            PingResult result = task.Result;
+        }
+        catch (AggregateException exception)
+        {
+            ExceptionDispatchInfo.Capture(exception.Flatten().InnerException!).Throw();
+        }
```
Comment (tnadalini) [inline Assignment/Assignment.Tests/PingProcessTests.cs:125]: not sure if the try catch is necessary, but otherwise this looks like a good test

**Review by tnadalini — COMMENTED** [review]
Implement PingProcess' public Task<PingResult> RunTaskAsync(string hostNameOrAddress) ✔
        First implement public void RunTaskAsync_Success() test method to test PingProcess.RunTaskAsync() using "localhost". ✔
        Do NOT use async/await in this implementation. ✔
    Implement PingProcess' async public Task<PingResult> RunAsync(string hostNameOrAddress) ✔
        First implement the public void RunAsync_UsingTaskReturn_Success() test method to test PingProcess.RunAsync() using "localhost" without using async/await. ✔
        Also implement the async public Task RunAsync_UsingTpl_Success() test method to test PingProcess.RunAsync() using "localhost" but this time DO using async/await. ✔
    Add support for an optional cancellation token to the PingProcess.RunAsync() signature. ✔ Inside the PingProcess.RunAsync() invoke the token's ThrowIfCancellationRequested() method so an exception is thrown. ✔ Test that, when cancelled from the test method, the exception thrown is an AggregateException ✔ with a TaskCanceledException inner exception ✔ using the test methods RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrapping ✔and RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrappingTaskCanceledException ✔ respectively.
    Complete/fix AND test async public Task<PingResult> RunAsync(IEnumerable<string> hostNameOrAddresses, CancellationToken cancellationToken = default) which executes ping for and array of hostNameOrAddresses (which can all be "localhost") in parallel, adding synchronization if needed. ✔ NOTE:
        The order of the items in the stdOutput is irrelevent and expected to be intermingled.
        StdOutput must have all the ping output returned (no lines can be missing) even though intermingled. ✔
    Implement AND test public Task<int> RunLongRunningAsync(ProcessStartInfo startInfo, Action<string?>? progressOutput, Action<string?>? progressError, CancellationToken token) using Task.Factory.StartNew() and invoking RunProcessInternal with a TaskCreation value of TaskCreationOptions.LongRunning and a TaskScheduler value of TaskScheduler.Current. Returning a Task<PingResult> is also okay. NOTE: This method does NOT use Task.Run.

Extra Credit

    Test and implement PingProcess.RunAsync(System.IProgress<T> progess) so that you can capture the output as it occurs rather than capturing the output only when the process finishes. ❌

Fundamentals

    Ensure you enable:
        nullable reference types is enabled ✔
        net6 targeted ✔
        C# 10.0 ✔
        and enabled .NET analyzers for both projects ✔
    All of the above should be unit tested ✔
    Choose simplicity over complexity ✔

**Comment by mmwoodfo** [comment]
Did you do a PR review? @mpennock5

**Pull request closed without merging**